packages/ui5-tooling-transpile/lib/task.js

```diff
@@ -31,54 +32,150 @@ module.exports = async function ({ workspace, dependencies, /* taskUtil,*/ optio
+				// we ignore d.ts files for transpiling
+				if (!resourcePath.endsWith(".d.ts")) {
+					// mark source for omit from build result
+					taskUtil.setTag(resource, taskUtil.STANDARD_TAGS.OmitFromBuildResult, true);
 
-				// create SourceMap resource if available
-				if (result.map) {
-					//add the transpiled file name
-					result.map.file = path.basename(filePath);
-					config.debug && log.info(`  + sourcemap ${filePath}.map`);
+					// transpile the source
+					config.debug && log.info(`Transpiling resource ${resourcePath}`);
+					const result = await babel.transformAsync(
+						source,
+						Object.assign({}, babelConfig, {
+							filename: resourcePath // necessary for source map <-> source assoc
+						})
+					);
 
-					const resourceMap = resourceFactory.createResource({
-						path: `${filePath}.map`,
-						string: JSON.stringify(result.map)
+					// create the ts file in the workspace
+					const transpiledResource = resourceFactory.createResource({
+						path: filePath,
+						string: normalizeLineFeeds(result.code)
 					});
+					workspace.write(transpiledResource);
+
+					// create sourcemap resource if available
+					if (result.map) {
+						result.map.file = path.basename(filePath);
+						config.debug && log.info(`  + sourcemap ${filePath}.map`);
 
-					workspace.write(resourceMap);
+						const resourceMap = resourceFactory.createResource({
+							path: `${filePath}.map`,
+							string: JSON.stringify(result.map)
+						});
+
+						workspace.write(resourceMap);
+					}
 				}
 			}
 		})
 	);
+
+	// generate the dts files for the ts files
+	if (config.transpileTypeScript && !config.skipGenerateDTS) {
+		try {
+			// dynamically require typescript
+			const ts = require("typescript");
+
+			// options to generate d.ts files only
+			const options = {
+				allowJs: true,
+				declaration: true,
+				emitDeclarationOnly: true
+				//traceResolution: true,
+			};
+
+			// Create a Program with an in-memory emit
+			const host = ts.createCompilerHost(options);
+			(host.getCurrentDirectory = function () {
+				return "";
+			}),
+				(host.fileExists = function (file) {
+					if (sourcesMap[file]) {
+						return true;
+					} else {
+						return fs.existsSync(file);
+					}
+				});
+			host.readFile = function (file) {
+				if (sourcesMap[file]) {
+					return sourcesMap[file];
+				} else {
+					return fs.readFileSync(file, "utf-8");
+				}
+			};
+			host.writeFile = function (fileName, content) {
+				config.debug && log.info(`Generating d.ts for resource ${fileName}`);
+				const dtsFile = resourceFactory.createResource({
+					path: `${fileName}`,
+					string: content
+				});
+				workspace.write(dtsFile);
+			};
+			host.resolveModuleNames = function (moduleNames, containingFile) {
+				const resolvedModules = [];
+				for (const moduleName of moduleNames) {
+					// try to use standard resolution
+					let result = ts.resolveModuleName(moduleName, containingFile, options, {
```

Comment:
can be `const`

Reply:
copy paste from here: https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API - not reworked all places properly... 😁